docs/closure-compiler.md

```diff
@@ -0,0 +1,697 @@
+      removeClass: (className) => this.root_.classList.remove(className),
+    });
+  }
+}
+```
+
+#### @typedefs are always `let` declarations, always pascal case, and always end in `Type`
+
+```js
+// GOOD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+let EventDataType;
+
+// GOOD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+export let EventDataType;
+
+// BAD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+MDCComponentFoundation.EventDataType;
+
+// BAD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+let eventDataType;
+
+// BAD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+let EventData;
+```
+
+Using this convention allows us to write tooling around handling these expressions, such as
+lint rule exceptions, and (in the future) code removal tools.
+
+### Objects that use event names or other external symbols as keys must be declared as `@dict` or `!Object<string, T>`.
+
+```js
+// GOOD
+/** @const {!Object<string, string>}  */
+const activationDeactivationPairs = {
+  'mousedown': 'mouseup',
+  'touchstart': 'touchend',
+};
+
+// GOOD
+/** @dict */
+window.settings = {
+  'windowObject': window,
+  'domReadyEvent': 'onready',
+  'scriptExecutionTimeoutMs': 3000,
+}
+
+// BAD (no quoted keys)
+/** @type {{mousedown: !Function, touchstart: !Function}} */
+const eventListenerMap = {
+  mousedown: (evt) => handleMouseup(evt),
+  touchstart: (evt) => handleTouchstart(evt),
+};
+```
+
+```js
+// GOOD
+el.addEventListener(activationDeactivationPairs['mousedown'], (evt) => this.deactivate(evt));
+
+// GOOD
+Object.keys(activationDeactivationPairs).forEach((activationEvt) => {
+  const deactivationEvt = activationDeactivationPairs[activationEvt];
+
+  el.addEventListener(activationEvt, (evt) => this.activate(evt));
+  el.addEventListener(deactivationEvt, (evt) => this.deactivate(evt));
+});
+
+// BAD
+el.addEventListener(activationDeactivationPairs.mousedown, (evt) => this.deactivate(evt));
+```
+
+By default, when closure uses advanced compilation, it rewrites the property names of objects to be
+as short as possible, ensuring the smallest possible code size. This is problematic when object
+properties have semantic meaning for code used outside of closure. For example, if object keys
+represent event names to be passed to `addEventListener`, or global settings to be affected by the
+user, then the code will break when closure rewrites the property names. In order to prevent this,
+objects with semantic keys must be declared as described above. Furthermore:
+
+- All object keys _must be quoted_
+- All references to object keys _must be done using bracket notation_
+
+#### `Object<string, T> vs. @dict`
+
+- Use `Object<string, T>` for objects where _the type for every must be the same_.
+- Use `@dict` for objects where _the type for every value can vary_.
+
+### An example component skeleton
+
+The following shows a set of skeleton files for an example component:
```

Comment:
Hmm I haven't. Would separating these into different files make it easier such that users could simply copy/paste the file over? Or copy the files into a new directory like a template? I'd argue that having them in one markdown file actually might be better for documentation purposes since it keeps all the code local to one specific file. Therefore rather than have to jump around to multiple different files users could use this one page as a reference. WDYT?